File: ruby/day1/captcha.rb
require 'minitest/autorun'

def captcha(num)
  sum = 0

  numbers = num.to_s.split('').map(&:to_i)

  size = numbers.size

  numbers.each_with_index do |n, i|
    possible_match = numbers[(i+1)%size]
    if n == possible_match
      sum += n
    end
  end

  return sum
end

def captcha2(num)
  sum = 0

  numbers = num.to_s.split('').map(&:to_i)

  size   = numbers.size
  offset = size/2

  numbers.each_with_index do |n, i|
    possible_match = numbers[(i+offset)%size]
    if n == possible_match
      sum += n
    end
  end

  return sum
end

class CaptchaTest < Minitest::Test
  def test_part_one
    assert_equal 3, captcha(1122)
    assert_equal 4, captcha(1111)
    assert_equal 0, captcha(1234)
    assert_equal 9, captcha(91212129)
  end

  def test_part_two
    assert_equal 6,  captcha2(1212)
    assert_equal 0,  captcha2(1221)
    assert_equal 4,  captcha2(123425)
    assert_equal 12, captcha2(123123)
    assert_equal 4,  captcha2(12131415)
  end
end

input = File.readlines('input.txt').first.to_i

puts "Part 1:"
puts captcha(input)

puts "Part 2:"
puts captcha2(input)

File: elixir/day1/captcha.ex
captcha = fn (array_of_numbers, calc_index) ->
  size = Enum.count(array_of_numbers)

  array_of_numbers
  |> Enum.with_index
  |> Enum.reduce(0, fn ({num, index}, acc) ->

    next_index = calc_index.(index, size)
    next       = Enum.at(array_of_numbers, next_index)

    case num == next do
      true  -> acc + num
      false -> acc
    end
  end)
end

{:ok, file} = File.open "input.txt"

# it'd be much better to read in by character, but hey
numbers =
  IO.read(file, :all)
  |> String.trim
  |> String.split("", [trim: true ])
  |> Enum.map(fn(x) -> String.to_integer(x) end)

IO.puts "Part 1: #{captcha.(numbers, fn (index, size) ->
  rem(index+1, size)
end)}"
IO.puts "Part 2: #{captcha.(numbers, fn (index, size) ->
  rem(index+(div(size,2)), size)
end)}"
